x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -0,0 +1,376 @@
+        Map<String, List<String>> removedNodeModelLookUp = new HashMap<>();
+        // TODO: make more efficient, right now this is O(nm) where n = sizeof(models) and m = sizeof(nodes)
+        // It could probably be O(max(n, m))
+        // Add nodes and keep track of currently routed nodes
+        // TODO: add memory aware node assignments.
+        // Should we indicate a partial allocation somehow if some nodes don't have space?
+        for (Map.Entry<String, TrainedModelAllocation> modelAllocationEntry : previousState.modelAllocations().entrySet()) {
+            for (DiscoveryNode node : currentState.getNodes()) {
+                // Only add the route if the node is NOT shutting down, this would be a weird case of the node
+                // just being added to the cluster and immediately shutting down...
+                if (isNodeShuttingDown(currentState, node.getId()) == false && modelAllocationEntry.getValue().canAllocateToNode(node)) {
+                    nodeHasCapacity(currentState, modelAllocationEntry.getValue().getTaskParams(), node).ifPresentOrElse(
+                        (error) -> builder.addFailedNode(modelAllocationEntry.getKey(), node.getId(), error),
+                        () -> builder.addNode(modelAllocationEntry.getKey(), node.getId())
+                    );
+                }
+            }
+            for (String nodeId : modelAllocationEntry.getValue().getNodeRoutingTable().keySet()) {
+                removedNodeModelLookUp.computeIfAbsent(nodeId, k -> new ArrayList<>()).add(modelAllocationEntry.getKey());
+            }
+        }
+
+        // Remove nodes
+        currentState.getNodes()
+            .forEach(
+                d -> {
+                    // If a node is referenced in the current state, we shouldn't remove the node
+                    // But, if that node that is referenced is shutting down, we should remove the node
+                    if (isNodeShuttingDown(currentState, d.getId()) == false) {
+                        removedNodeModelLookUp.remove(d.getId());
+                    }
+                }
+            );
+        for (Map.Entry<String, List<String>> nodeToModels : removedNodeModelLookUp.entrySet()) {
+            final String nodeId = nodeToModels.getKey();
+            for (String modelId : nodeToModels.getValue()) {
+                builder.removeNode(modelId, nodeId);
+            }
+        }
+        return update(currentState, builder);
+    }
+
+    static boolean shouldAllocateModels(final ClusterChangedEvent event) {
+        // If there are no allocations created at all, there is nothing to update
+        final TrainedModelAllocationMetadata newMetadata = event.state().getMetadata().custom(TrainedModelAllocationMetadata.NAME);
+        if (newMetadata == null) {
+            return false;
+        }
+        boolean masterChanged = event.previousState().nodes().isLocalNodeElectedMaster() == false;
+        if (event.nodesChanged() || masterChanged) {
```

Comment:
We only attempt to update the routing table for allocations nodes if:

 - nodes added
 - nodes removed
 - master node changed

The reason for the last point is that the local node itself (master) may also be a (ml) node and thus has the model assigned here. While this is not desired in ESS, its conceivable (especially in our integration testing framework)

Reply:
I am not sure I understand why we need to allocate models when the master node changed.

If the node that became master was in the cluster previously, then if it was also an ML node, it would have a routing for the allocation already. If the node is new, then there should also be a node added event so we'll get a chance to add a routing then. Am I missing something?

Reply:
@dimitris-athanasiou good points, I can remove the master node check.